File: elixir/lib/aoc/day2.ex
defmodule Aoc.Day2 do
  @input Path.join([__DIR__, "..", "..", "..", "inputs", "day2.txt"])

  def run_part_1 do
    read_input()
    |> Stream.filter(fn {password, {occurrences_rule, char}} ->
      password
      |> String.graphemes()
      |> Enum.count(&(&1 == char))
      |> Kernel.in(occurrences_rule)
    end)
    |> Enum.count()
    |> IO.puts()
  end

  def run_part_2 do
    read_input()
    |> Stream.filter(fn {password, {occurrences_rule, char}} ->
      [
        String.at(password, occurrences_rule.first - 1),
        String.at(password, occurrences_rule.last - 1)
      ]
      |> Enum.count(&(&1 == char))
      |> Kernel.==(1)
    end)
    |> Enum.count()
    |> IO.puts()
  end

  defp read_input do
    @input
    |> File.stream!()
    |> Stream.map(&String.trim/1)
    |> Stream.reject(&(&1 == ""))
    |> Stream.map(&parse_password_and_rules/1)
  end

  @line_regex ~r/(?<min>\d+)-(?<max>\d+) (?<char>.): (?<password>.+)/

  defp parse_password_and_rules(line) do
    %{"min" => min, "max" => max, "char" => char, "password" => password} =
      Regex.named_captures(@line_regex, line)

    {password, {String.to_integer(min)..String.to_integer(max), char}}
  end
end

File: rust/src/aoc/day2.rs
use regex::Regex;
use std::fs;
use std::io;
use std::io::prelude::*;
use std::str::FromStr;

#[derive(Debug)]
struct Range {
    first: usize,
    last: usize,
}

#[derive(Debug)]
struct PasswordAndRules {
    password: String,
    rule: Range,
    character: char,
}

lazy_static! {
    static ref LINE_REGEX: Regex =
        Regex::new(r"(?P<min>\d+)-(?P<max>\d+) (?P<char>.): (?P<password>.+)").unwrap();
}

impl FromStr for PasswordAndRules {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let captures = LINE_REGEX.captures(s).unwrap();

        Ok(PasswordAndRules {
            password: captures["password"].to_string(),
            rule: Range {
                first: captures["min"].parse().unwrap(),
                last: captures["max"].parse().unwrap(),
            },
            character: captures["char"].chars().next().unwrap(),
        })
    }
}

const INPUT: &str = "../inputs/day2.txt";

fn read_input() -> impl Iterator<Item = PasswordAndRules> {
    let input = fs::File::open(INPUT).unwrap();
    let file_contents = io::BufReader::new(input);

    file_contents.lines().map(|s| s.unwrap().parse().unwrap())
}

#[allow(dead_code)]
pub fn run_part_1() {
    let res: usize = read_input()
        .filter(|password_and_rules| {
            let char_count = password_and_rules
                .password
                .chars()
                .filter(|c| *c == password_and_rules.character)
                .count();

            char_count >= password_and_rules.rule.first
                && char_count <= password_and_rules.rule.last
        })
        .count();

    println!("{:?}", res);
}

#[allow(dead_code)]
pub fn run_part_2() {
    let res: usize = read_input()
        .filter(|password_and_rules| {
            let chars: Vec<_> = password_and_rules.password.chars().collect();

            let char_count = vec![password_and_rules.rule.first, password_and_rules.rule.last]
                .iter()
                .map(|n| chars[n - 1])
                .filter(|c| *c == password_and_rules.character)
                .count();

            char_count == 1
        })
        .count();

    println!("{:?}", res);
}
